app/jobs/scrape_keyword_job.rb

```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class ScrapeKeywordJob < ApplicationJob
+  queue_as :default
+
+  sidekiq_options retry: 0
+
+  retry_on GoogleScraperRuby::Errors::ScrapeError, wait: 1.minute, attempts: 3 do |job, _error|
+    keyword_id = job.arguments.first[:keyword_id]
+    # TODO(Goose97): better handle database failure
+    # If the below update crashes, the system ends up in an inconsistent state: the job still
+    # has :processing status but already moved to Sidekiq dead queue
+    Keyword.update(keyword_id, status: :failed)
```

Comment:
how about we set it as failed right in the `perform` action 🤔 

```
def perform(keyword_id:)
    Keyword.update(keyword_id, status: :processing)
    Google::ScrapeService.new(keyword_id: keyword_id).call!
  rescue GoogleScraperRuby::Errors::ScrapeError => error
    Keyword.update(keyword_id, status: :failed)
    raise error unless error.kind == :invalid_keyword
  end
```

Reply:
Oh, I see your point. So the job will be seen as `:failed` in between retry attempts, and mark as `:processing` before each attempt.

My current goal is: only marks a job as `:failed ` if it exceeds the maximum retries count (which is 3). This is not an AC so I guess we can relax that.

But even if we go with your suggestion, the issue is still there, am I right? If the update line crashes, there no retry mechanism left to catch the job and it go straight to Sidekiq's dead queue. Anw, this is not a serious issue, I think we can left it like this

Reply:
It is indeed not a major concern as the UI could have a retry mechanism (out of scope though for the IC). Always aim for covering the happy paths first. Further improvements for edge cases can be handled in subsequent stories.